Marlin/src/gcode/bedlevel/G35.cpp

```diff
@@ -178,11 +158,11 @@ void GcodeSuite::G35() {
   // the probe deployed if it was successful.
   probe.stow();
 
+  // Move XYZ to wait position
+  move_to_tramming_wait_pos();
+
   // After this operation the Z position needs correction
   set_axis_never_homed(Z_AXIS);
-
-  // Home Z after the alignment procedure
-  process_subcommands_now_P(PSTR("G28Z"));
```

Comment:
> 
> 
> > I would consider to make a change some lines above (beginning of the percedure) from home_all_axes(); to queue.inject_P(PSTR("G28O"));
> 
> With small changes on the screws the bed center could stay within the `Z_PROBE_LOW_POINT` area, but if the bed was badly tilted, it isn't guaranteed that the middle of the bed would stay within the `Z_PROBE_LOW_POINT` area. And we cannot be sure that homing position is at the center of the bed.
> So I don't think it would be a good idea to change the first home to an optional one, because if the previous bed adjustment after the G35 is too great, then the `Z_PROBE_LOW_POINT` would trigger and give use an error.
> Even if I'm wrong in this, I still think the first home should stay there...

I agree 👍 

And I stay with my opinion for the end of the precedure. I would like to see a rehoming as follow:

```
  // If probe is used for homing and there is no fixed endstop, rehome Z
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
    process_subcommands_now_P(PSTR("G28Z"));
  #endif

  // Move XYZ to wait position
  move_to_tramming_wait_pos();
```

Reply:
I still don't think that there is any reason to home after G35, because at the end the z axis will be set to be unhomed (`set_axis_never_homed(Z_AXIS)`).

Reply:
Well at least I think `set_axis_never_homed(Z_AXIS)` should only be done if `HOMING_Z_WITH_PROBE` is true. Or are there printers out there, which have the z endstop moving with bed adjustment up and down?

I think it would be more convenient to have it homed. Especially because it's Z only. But to have consistency over all tramming / level bed corners options I think it okey to just do:

```
  // If probe is used for homing meassuring the bed heigth, set Z unhomed
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
  #endif
```
 And i think we should add this also to `Marlin/src/lcd/menu/menu_tramming.cpp`

Reply:
> Well at least I think `set_axis_never_homed(Z_AXIS)` should only be done if `HOMING_Z_WITH_PROBE` is true. Or are there printers out there, which have the z endstop moving with bed adjustment up and down?

Not sure why should `set_axis_never_homed(Z_AXIS)` depend on `HOMING_Z_WITH_PROBE`. It only sets the axis untrasted and unhomed, so that the next homing must happen, but the stepper will still be engaged, at least I believe so.
I'm a bit lost on your second sentence, could you explain it a bit more?

> I think it would be more convenient to have it homed. Especially because it's Z only. But to have consistency over all tramming / level bed corners options I think it okey to just do:
> 
> ```
>   // If probe is used for homing meassuring the bed heigth, set Z unhomed
>   #if HOMING_Z_WITH_PROBE
>     set_axis_never_homed(Z_AXIS);
>   #endif
> ```
> 
> And i think we should add this also to `Marlin/src/lcd/menu/menu_tramming.cpp`

One of my problem with the homing at the end is that it isn't just a Z homing, but a full homing sequence, at least for me with a BLTouch, and it takes "way" too much time, and for what? After the bed adjustment we cannot trust the previously measured homing height, and that is why there is the `set_axis_never_homed(Z_AXIS)` at the end.

Reply:
> Not sure why should `set_axis_never_homed(Z_AXIS)` depend on `HOMING_Z_WITH_PROBE`. It only sets the axis untrasted and unhomed, so that the next homing must happen, but the stepper will still be engaged, at least I believe so.
> I'm a bit lost on your second sentence, could you explain it a bit more?

If we have homing with an endstop mounted to the gantry, there is no need to set Z untrusted because it's still at the same `current_position.z` relative to `Z_HOME_POS`. So it's only needed to go for `set_axis_never_homed(Z_AXIS);` if `HOMING_Z_WITH_PROBE`.

> One of my problem with the homing at the end is that it isn't just a Z homing, but a full homing sequence, at least for me with a BLTouch, and it takes "way" too much time, and for what? After the bed adjustment we cannot trust the previously measured homing height, and that is why there is the `set_axis_never_homed(Z_AXIS)` at the end.

`G28Z ` will home only Z even with BLTouch, because X and Y are known/trusted. I don't know your config, but am I right that you use an Ender3? (Same as me, also with BLTouch). My Printer needs max about 6 to 7 seconds to do G28 Z.

Reply:
> If we have homing with an endstop mounted to the gantry, there is no need to set Z untrusted because it's still at the same `current_position.z` relative to `Z_HOME_POS`. So it's only needed to go for `set_axis_never_homed(Z_AXIS);` if `HOMING_Z_WITH_PROBE`.

By my understanding `set_axis_never_homed(Z_AXIS)` isn't needed because the Z position changes relative to the endstop/probe, but the bed height changes, and by that the Z position relative to the bed. If the Z axis stays trusted at the end and the bed adjustment was relatively large, then on the next Z movement to zero, the nozzle would grind into the bed, which is why `set_axis_never_homed(Z_AXIS)` is needed for both endstop and probe.

> `G28Z ` will home only Z even with BLTouch, because X and Y are known/trusted. I don't know your config, but am I right that you use an Ender3? (Same as me, also with BLTouch). My Printer needs max about 6 to 7 seconds to do G28 Z.

You are right, `process_subcommands_now_P(PSTR("G28Z"))` does only move to `Z_SAFE_HOMING` position then only homes Z, but even if it would take less than a second I still think that it is just an unnecessary extra movement, as `set_axis_never_homed(Z_AXIS)` needs to be at the end, because of what is written above.

Reply:
> By my understanding `set_axis_never_homed(Z_AXIS)` isn't needed because the Z position changes relative to the endstop/probe, but the bed height changes, and by that the Z position relative to the bed.

**isn't?** typo?

Z position **only** changes relative to the bed, because the bed is moved/adjusted!

>If the Z axis stays trusted at the end and the bed adjustment was relatively large, then on the next Z movement to zero, the nozzle would grind into the bed, which is why `set_axis_never_homed(Z_AXIS)` is needed for both endstop and probe.

This is only valid for a probe/endstop that touches the bed for homing! And therefor `HOMING_Z_WITH_PROBE`.

Reply:
Unresolving this comment, since it seems important to see this discussion during review.

Reply:
> **isn't?** typo?

Sorry, it should be "is", edited the comment. Thank you for catching it.

> Z position **only** changes relative to the bed, because the bed is moved/adjusted!

Yes, that is the problem. ⬇️ 

> This is only valid for a probe/endstop that touches the bed for homing! And therefor `HOMING_Z_WITH_PROBE`.

If the bed would only be adjusted downwards, then it wouldn't be a problem, but that isn't always the case. If the bed is adjusted upwards, then the previously measured z position will be wrong and unusable and potentially dangerous if used.

Reply:
I agree that invalidating Z would only help if `HOMING_Z_WITH_PROBE`. If you are homing with the endstop it will return to the same position, because endstops don't measure the bed surface!

Reply:
I also agree that there is no need to actually home _anything_ after the G35 command. There are two possible outcomes from G35, and homing makes sense in neither of them.

1. No screw adjustment needed.

2. Screw adjustment needed, **but they aren't adjusted yet!**. There is no point to homing Z, before the screws are actually adjusted!

Reply:
@sjasonsmith Do you think this needs other opinions or can be set as resolved?

Reply:
> I agree that invalidating Z would only help if `HOMING_Z_WITH_PROBE`. If you are homing with the endstop it will return to the same position, because endstops don't measure the bed surface!

I think that it cannot hurt to have z invalidation on a z-endstop setup too and to know over time that it is useless or not, but it would be an easy fix:
```
  // If probe is used for homing meassuring the bed heigth, set Z never homed
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
  #endif
```

Reply:
I don’t feel too strongly about it. You are already removing the actual homing at the end of the routine. We can wait and see if people still complain.